### ISSUE TITLE
Controller.Module.Action Event is case-sensitive

### ISSUE THREAD

**Issue by Zeichen32** [piwik/piwik#8422]
I've noticed that Events triggered by the FrontController Dispatcher are case-sensitive, but the called actions are is case-insensitive.

For example:
?module=CoreAdminHome&action=optout => triggers action CoreAdminHome:***optOut***
?module=CoreAdminHome&action=optOut => triggers action CoreAdminHome:***optOut***
?module=CoreAdminHome&action=OptOut => triggers action CoreAdminHome:***optOut***

Events:
?module=CoreAdminHome&action=optout => triggers event Controller.CoreAdminHome.***optout***
?module=CoreAdminHome&action=optOut => triggers event Controller.CoreAdminHome.***optOut***
?module=CoreAdminHome&action=OptOut => triggers event Controller.CoreAdminHome.***OptOut***

So the docs or the behavior is wrong:

```
* Triggered directly before controller actions are dispatched.
*
* This event exists for convenience and is triggered directly after the {@hook Request.dispatch}
* event is triggered.
*
* It can be used to do the same things as the {@hook Request.dispatch} event, but for one controller
* action only. Using this event will result in a little less code than {@hook Request.dispatch}
```

File: https://github.com/piwik/piwik/blob/master/core/FrontController.php#L492

**Comment by tsteur** [comment]
I'm a bit surprised to are all rendered perfectly. We should always use the correct action name as defined in the code or alternatively always lowercase I don't really mind :) We should have a look why it renders all of those actions when only one should be valid

**Comment by Zeichen32** [comment]
Its because php functions are case-insensitive.

```
Note: Function names are case-insensitive, though it is usually good form to call functions as they appear in their declaration.
```
http://php.net/manual/en/functions.user-defined.php

The easiest way were to convert the controller action name to lowercase. But it's maybe a BC.

**Comment by mattab** [comment]
I would suggest that `&action=` should match the controller action as found in code (case sensitive) VS using lowercase only

**Comment by Zeichen32** [comment]
If we change the ControllerResolver it can be done case-sensitive:

https://github.com/piwik/piwik/blob/master/core/Http/ControllerResolver.php#L74
```php
/** @var $controller Controller */
$controller = $this->abstractFactory->make($controllerClass);

$action = $action ?: $controller->getDefaultAction();

if (!is_callable(array($controller, $action)) || !in_array($action, get_class_methods($controller))) {
    return null;
}
```

**Comment by tsteur** [comment]
:+1: sounds good, should also not cause any performance issues

**Comment by Zeichen32** [comment]
Ive create a PR to fix this issue #8426

**Comment by tsteur** [comment]
We could afterwards write a test that gets a list of all events (starting with `Controller.`) of all plugins and check whether the controller actions actually exist. This way we make sure we don't break anything and will get notified easily in case one has a typo in the future

**Comment by Zeichen32** [comment]
I will add some tests

**Comment by tsteur** [comment]
This issue was closed via #8426

**Comment by tsteur** [comment]
From my understanding this issue was fixed
